--- a/system/etc/init/hw/init.rc
+++ b/system/etc/init/hw/init.rc
@@ -2,6 +2,9 @@
 import /init.recovery.samsung.rc
 
 on early-init
+    write /sys/fs/selinux/enforce 0
+    exec u:r:recovery:s0 -- /system/bin/setenforce 0
+
     # Set the security context of /postinstall if present.
     restorecon /postinstall
 
@@ -53,6 +56,10 @@
    rm /dev/.booting
 
 # Mount filesystems and start core system services.
+on post-fs
+    # lpmapped: create /dev/block/mapper/* before recovery service
+    exec u:r:recovery:s0 root root -- /system/bin/lpmode-run
+
 on late-init
     trigger early-fs
     trigger fs
